### DISCUSSION TITLE
Unable to Integrate webkit2gtk-5.0 with Relm4 app targeting GTK4 on Ubuntu 22.04 LTS

### DISCUSSION THREAD

**Discussion by cdspacedos** [Relm4/Relm4#501] [Q&A] [answered]
I'm building an application using the latest release of Relm4 for GTK4 on Ubuntu 22.04 LTS and I need to embed a web browser component for 3rd party authentication. Previously with GTK3, I used webkit2gtk-4.0.

Now, aiming for GTK4 support, I tried to integrate webkit2gtk-5.0 which is GTK4-based. However, I'm facing version conflicts with gtk4-sys between Relm4 and the webkit project. I also gave webkitgtk-6.0 a shot but ran into similar issues.

Has anyone successfully embedded a GTK4-compatible web component in a Relm4 application? Any guidance or workarounds would be appreciated.

**Comment by AaronErhardt** [comment]
Can you show me the error message? Also, you could try to use `gtk4` with version 0.6 instead of Relm4 to test if that's an issue with Relm4 or another crate.

**Reply by cdspacedos** [reply]
Thank you for the quick reply!

I don't think this is a fault of the Relm4, but it would be very useful to find the way around the dependency conflict, as currently I don't see the way to have the embedded browser in Relm4 app.

With dependencies below, program compiles, but has no browser functionality:

```rust
[dependencies]
gtk = { version = "0.6.6", package = "gtk4" }
relm4 = "0.6"
relm4-macros = "0.6"
relm4-components = "0.6"
```

In attempt to add the webkit6 to dependencies, I'm getting the following error:
```cargo add webkit6
    Updating crates.io index
      Adding webkit6 v0.2.0 to dependencies.
             Features:
             - v2_42
    Updating crates.io index
error: failed to select a version for `gtk4-sys`.
    ... required by package `gtk4 v0.6.6`
    ... which satisfies dependency `gtk = "^0.6.6"` of package `icarus v0.1.0 (/home/miljkovi/Temp/relm/Icarus)`
versions that meet the requirements `^0.6` are: 0.6.3, 0.6.2, 0.6.0

the package `gtk4-sys` links to the native library `gtk-4`, but it conflicts with a previous package which links to `gtk-4` as well:
package `gtk4-sys v0.7.0`
    ... which satisfies dependency `ffi = "^0.7"` of package `gtk4 v0.7.1`
    ... which satisfies dependency `gtk = "^0.7"` of package `webkit6 v0.2.0`
    ... which satisfies dependency `webkit6 = "^0.2.0"` of package `icarus v0.1.0 (/home/miljkovi/Temp/relm/Icarus)`
Only one package in the dependency graph may specify the same links value. This helps ensure that only one copy of a native library is linked in the final binary. Try to adjust your dependencies so that only one package uses the links ='gtk4-sys' value. For more information, see https://doc.rust-lang.org/cargo/reference/resolver.html#links.

failed to select a version for `gtk4-sys` which could resolve this conflict
```

I did a little experiment with gtk4-rs and webkit6, and was able to make it working on Ubuntu 22.04.

libwebkit2gtk-5.0 is not available on Ubuntu 22.04. Compiling the webkit2gtk5 from source and including it to the dependencies causes the conflict similar to the one described above.

**Reply by AaronErhardt** [reply, marked as answer]
Should be easy to fix. You need to use version 0.1 of webkit6 because 0.2 already uses gtk-rs 0.7. Relm4 follows the version numbers of its underlying gtk-rs version btw. so Relm4 0.6 uses gtk-rs 0.6.

**Reply by cdspacedos** [reply]
Thank you Aaron. Changing the webkit version to 0.1.0 did the trick. I can compile the code now!

Can you point me to the code sample (if any) of integrating the standard gtk4-rs components with Relm4?

**Reply by AaronErhardt** [reply]
> Can you point me to the code sample (if any) of integrating the standard gtk4-rs components with Relm4?

Can you be more precise? Technically, gtk4-rs has no components, but widgets whereas Relm4 allows you to combine widgets into components, but you can't create widgets with Relm4.

**Reply by cdspacedos** [reply]
Understood. I will go back to coding now, and will ask separate question if necessary.
